### ISSUE TITLE
Defining envmap as optimizable gives different result

### ISSUE THREAD

**Issue by dorverbin** [mitsuba-renderer/mitsuba3#417]
## Summary

Rendering a scene with an envmap and then making it optimizable and setting it to the ground truth value gives a different result.

## System configuration

System information:

  OS: CentOS Linux release 7.9.2009 (Core)
  CPU: Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
  GPU: Tesla V100-SXM2-16GB
  Python: 3.8.13 (default, Mar 28 2022, 11:38:47) [GCC 7.5.0]
  NVidia driver: 515.65.01
  CUDA: 9.0.176
  LLVM: 0.0.0

  Dr.Jit: 0.2.2
  Mitsuba: 3.0.2
     Is custom build? False
     Compiled with: GNU 10.2.1
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb

## Description

Rendering a simple sphere with the default `conductor` BSDF gives significantly different results before and after setting the optimizer to optimize the envmap. The errors, often in the order of ~0.1-0.2, seem to be mostly at the poles, clearly visible at the center of the image (see attached difference image, multiplied by ~5).
![mitsuba_envmap_opt](https://user-images.githubusercontent.com/15837806/204960608-db270293-3db7-4af8-aeb0-11c1a3cbc0dd.png)

## Steps to reproduce

Minimal example creating a random envmap, defining a scene, and rendering twice (with and without the optimizer):

```
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
import mitsuba as mi
mi.set_variant('cuda_ad_rgb')

# Create random (but azimuthally-constant) envmap.
envmap = np.uint8(255.0 * np.random.rand(50, 1, 3)).repeat(100, 1)
Image.fromarray(envmap).save('temp.png')

# Create simple spherical scene.
scene_dict = {
    'type': 'scene',
    'integrator': {
        'type': 'path',
        'max_depth': 2,
        'hide_emitters': True
    },
    'sensor': {
        'type': 'perspective',
        'fov': 39.6,
        'near_clip': 0.1,
        'far_clip': 1000.0,
        'to_world': mi.ScalarTransform4f.look_at(
            origin=[0, 4, 0],
            target=[0, 0, 0],
            up=[0, 0, 1]
        ),
        'sampler': {
            'type': 'independent',
            'sample_count': 2048
        },
        'film': {
            'type': 'hdrfilm',
            'width': 64,
            'height': 64,
            'pixel_format': 'rgb'
        }
    },
    'envmap': {
        'type': 'envmap',
        'filename': 'temp.png'
    },
    'sphere': {
        'type': 'sphere',
        'bsdf': {
            'type': 'conductor'
        }
    }
}

# Use 1 spp, although any number behaves similarly.
spp = 1
seed = 0

# Load scene and render.
scene = mi.load_dict(scene_dict)
image_ref = mi.render(scene, spp=spp, seed=seed)

# Get params and define optimizer.
params = mi.traverse(scene)
opt = mi.ad.SGD(lr=0.1)
opt['envmap.data'] = params['envmap.data']
params.update(opt)

# Render again.
image = mi.render(scene, params, spp=spp, seed=seed)

# Compute the difference image, print min/max, and plot.
diff = (image - image_ref).numpy()
print(diff.min(), diff.max())
normalized_diff = 0.5 * diff / max([abs(diff.min()), abs(diff.max())]) + 0.5
plt.imshow(normalized_diff)
plt.axis('off')
```

**Comment by njroussel** [comment]
Hi @dorverbin 

Thanks for the reproducer. I was able to reproduce the issue and I've already somewhat isolated the cause.
The difference comes from the call to `params.update`. This is notifying the `envmap` that it's data has changed and it should recompute it's internal data structures which are, namely, used to sample it.

What's suprising here is that these data structures should, even though they're recomputed, stay the same as the data did not actually change. In addition, it seems that the differences are not deterministic :exploding_head:. Individual runs produce different outputs. 

I'll keep digging and post any further updates.

**Comment by njroussel** [comment]
This has been fixed in #440. It should be available in next public release (later today).
Thank you for the reproducer.
